### DISCUSSION TITLE
Changed edge_label_index after LinkNeighborLoader

### DISCUSSION THREAD

**Discussion by songsong0425** [pyg-team/pytorch_geometric#7793] [Q&A] [answered]
Before starting, sorry for the frequent question.
While I running the `LinkNeighborLoader` code, I found that the `edge_label_index` was changed.
For example, I have a graph dataset and split it into the train, validation, and test dataset. (For simplicity, I only consider the training dataset)
```python
train_data
# Data(edge_index=[2, 2159112], pos_edge_label=[343], pos_edge_label_index=[2, 343], neg_edge_label=[343], neg_edge_label_index=[2, 343], edge_label=[686], edge_label_index=[2, 686], x=[47822, 259])
train_data.edge_label_index
# tensor([[ 5296,   613,   322,  ..., 19970, 24669, 27129],
#         [18473,  9092,  9092,  ...,  9153, 12671, 19081]])
```

When I run the `LinkNeighborLoader`, the `edge_label_index` is changed to the below format:
```python
train_loader = LinkNeighborLoader(train_data, edge_label_index=train_data.edge_label_index, batch_size=3, shuffle=True, neg_sampling_ratio=1.0, num_neighbors=[3,3])

for data in train_loader:
    print(data.edge_label_index)
# tensor([[ 1,  3,  5,  4,  7,  6],
#         [ 2,  0, 10,  8, 11,  9]])
# tensor([[ 0,  2,  6, 10,  1, 11],
#         [ 9,  3,  5,  7,  8,  4]])
# ...
```

Although I confirmed that the model learned the features and read the PyG documentation (https://pytorch-geometric.readthedocs.io/en/latest/advanced/batching.html), I cannot understand why the range of the edge index is changed.
If I performed the wrong process, please tell me where should I fix it.

**Comment by akihironitta** [comment, marked as answer]
Hey @songsong0425! I reported this issue via PyG Slack last night, and #7791 has just fixed the issue!

**Reply by songsong0425** [reply]
Oh, sorry for the duplicated question. I didn't know it because I'm not in the PyG Slack. Thank you for announcing. Besides, what should I do for using the fixed version?

**Reply by EdisonLeeeee** [reply]
@songsong0425 You can install the nightly version of PyG: https://github.com/pyg-team/pytorch_geometric#nightly-and-master

**Reply by songsong0425** [reply]
Thank you for your kind reply. Since it's my first time using the nightly version, I'm not quite sure. Can I use it just like the regular PyTorch Geometric? (i.e., `from torch_geometric.loader import LinkNeighborLoader` ...)

**Reply by EdisonLeeeee** [reply]
Sure :)

**Reply by songsong0425** [reply]
Uhm... But it still returns the same result. I tried to run the code after installing the nightly build and restarting the kernel in the jupyter notebook.

**Reply by EdisonLeeeee** [reply]
Weird. How did you install PyG? You might need to uninstall it first:
```bash
pip uninstall torch_geometric -y
pip install git+https://github.com/pyg-team/pytorch_geometric.git
```

**Reply by songsong0425** [reply]
I installed PyG nightly using `pip install pyg-nightly` which is described on the main page of the PyG GitHub. Please see the attached image that was installed in my virtual environment. I'll try to run your code. Thanks a lot!

from
![image](https://github.com/pyg-team/pytorch_geometric/assets/80442548/76c26fbf-e941-4a08-a77a-28616b799c17)
![image](https://github.com/pyg-team/pytorch_geometric/assets/80442548/099bd96f-651e-410f-a4da-22d5d509da2d)

to

![image](https://github.com/pyg-team/pytorch_geometric/assets/80442548/12bf4f11-5e7e-4fb0-aab2-a6e4b4cd6c3b)
![image](https://github.com/pyg-team/pytorch_geometric/assets/80442548/a468da80-f9a9-4524-9991-10bd59ed089f)

Unfortunately, I couldn't get the corrected result... I'll try again.
+ I got a new warning after updating the library: `/home/(blind username)/anaconda3/envs/PyG/lib/python3.9/site-packages/torch_geometric/typing.py:65: UserWarning: An issue occurred while importing 'torch-spline-conv'. Disabling its usage. Stacktrace: /lib/x86_64-linux-gnu/libm.so.6: version GLIBC_2.27 not found (required by /home/songyeon/anaconda3/envs/PyG/lib/python3.9/site-packages/torch_spline_conv/_basis_cuda.so)
  warnings.warn(`. Is it the critical warning for running the code? If not, please ignore it. Thanks!

**Reply by EdisonLeeeee** [reply]
Could you please try running this and let me know what it prints?
```bash
python -c "import torch_geometric; print(torch_geometric.__version__); print(torch_geometric)"
```

**Reply by songsong0425** [reply]
Sure! Here you are.
![image](https://github.com/pyg-team/pytorch_geometric/assets/80442548/3f890faf-4aa3-4455-bc7b-c1f480cc3fa3)

(jupyter notebook)
![image](https://github.com/pyg-team/pytorch_geometric/assets/80442548/abac4c2b-117f-49e0-aa5c-1dff04d2b358)

**Reply by EdisonLeeeee** [reply]
Oh, I see. Maybe we misunderstood your question. Yes, the range of the edge indices is changed since they are relabeled in each batch. This is an intended behavior :)

**Reply by songsong0425** [reply]
Wow, now I understand that it was intentional. Could you let me know what has been fixed this time? I'm curious to know how it differs from my previous question.

**Reply by EdisonLeeeee** [reply]
Feel free to join our slack and have a look at: https://torchgeometricco.slack.com/archives/C057729E6G2/p1690145073354439

**Reply by songsong0425** [reply]
Thank you for announcing the Slack channel. I'll explore it for more information. Once again, thank you very much for your help!

**Reply by EdisonLeeeee** [reply]
You're welcome. If you have any questions or need further assistance, don't hesitate to reach out. We're always here to help :)

**Reply by pstrybol** [reply]
Hi @EdisonLeeeee , I also noticed this recently but I fail to see how the new labels are mapped? The `node_id` and `n_id` are still the same as in the original `HeterData` object?

**Reply by EdisonLeeeee** [reply]
Hi @pstrybol , the node index in a batch is remapped to a new index and currently, we do not provide the mapping from the original index to the new one. For `n_id`, it represents the global node index of each node in the batch, i.e., the original node index.

**Reply by pstrybol** [reply]
Thanks for the reply! The reason I would need the node mapping is because I want to calculate the correlation between the predicted probabilities and original scores but I guess I could just input custom `edge_index` after each epoch for example, in `eval()` mode so it does not update weights? I also had some more question around `LinkNeighborLoader` but I asked them in a separate discussion topic #7797.
Cheers
